CSharpSourceCode/HarmonyPatches/CampaignPatches.cs

```diff
@@ -60,55 +58,108 @@ public static bool Prefix(string id, MBObjectManager __instance)
             else return true;
         }
 
-        //Change vanilla behaviour. Spawn every type of companion/wanderer at every culture matching town.
+        [HarmonyPostfix]
+        [HarmonyPatch(typeof(UrbanCharactersCampaignBehavior), "OnNewGameCreated")]
+        public static void AfterWandererTemplatesBuilt(ref List<CharacterObject> ____companionTemplates)
+        {
+            ____companionTemplates = ____companionTemplates.Where(x => x.IsTOWTemplate()).ToList();
+        }
+
+        [HarmonyPostfix]
+        [HarmonyPatch(typeof(UrbanCharactersCampaignBehavior), "OnGameLoaded")]
+        public static void AfterWandererTemplatesBuilt2(ref List<CharacterObject> ____companionTemplates)
+        {
+            ____companionTemplates = ____companionTemplates.Where(x => x.IsTOWTemplate()).ToList();
+        }
+
         [HarmonyPrefix]
-        [HarmonyPatch(typeof(UrbanCharactersCampaignBehavior), "CreateCompanion")]
-        public static bool Prefix2(CharacterObject companionTemplate, UrbanCharactersCampaignBehavior __instance, ref List<Hero> ____companions)
+        [HarmonyPatch(typeof(UrbanCharactersCampaignBehavior), "SpawnUrbanCharactersAtGameStart")]
+        public static bool SpawnWanderersAtStart(UrbanCharactersCampaignBehavior __instance, ref List<Hero> ____companions, List<CharacterObject> ____companionTemplates)
+        {
+            List<CharacterObject> list = ____companionTemplates.Where(x => x.IsTOWTemplate()).ToList();
+            foreach (Settlement settlement in Settlement.All)
+            {
+                if (settlement.IsTown)
+                {
+                    int targetNotableCountForSettlement = Campaign.Current.Models.NotableSpawnModel.GetTargetNotableCountForSettlement(settlement, Occupation.Artisan);
+                    for (int i = 0; i < targetNotableCountForSettlement; i++)
```

Comment:
I am mostly a man of a for each loop, but I expierence an issue with that game that this sometimes for some reason not work. did I miss here something, here again I see the classic for loop instead of the foreach loop

Reply:
foreach would do this - well, for each, but there is a different use case here, the goal is to only get as many as the notablespawnmodel defines, even if the list actually contains more entries.